apps/web/lib/trpc/routers/keySettings.ts

```diff
@@ -0,0 +1,409 @@
+        ratelimitRefillRate: z.number().int().positive().optional(),
+        ratelimitRefillInterval: z.number().int().positive().optional(),
+      }),
+    )
+    .mutation(async ({ input, ctx }) => {
+      let ratelimitType: "fast" | "consistent" | null = null;
+      let ratelimitLimit: number | null = null;
+      let ratelimitRefillRate: number | null = null;
+      let ratelimitRefillInterval: number | null = null;
+
+      if (input.enabled) {
+        if (typeof input.ratelimitType !== "string") {
+          throw new TRPCError({
+            message: "ratelimitType must be a string",
+            code: "BAD_REQUEST",
+          });
+        }
+        ratelimitType = input.ratelimitType;
+
+        if (typeof input.ratelimitLimit !== "number" || input.ratelimitLimit <= 0) {
+          throw new TRPCError({
+            message: "Limit must be a positive integer",
+            code: "BAD_REQUEST",
+          });
+        }
+        ratelimitLimit = input.ratelimitLimit;
+
+        if (typeof input.ratelimitRefillRate !== "number" || input.ratelimitRefillRate <= 0) {
+          throw new TRPCError({
+            message: "Rate must be a positive integer",
+            code: "BAD_REQUEST",
+          });
+        }
+        ratelimitRefillRate = input.ratelimitRefillRate;
+        if (
+          typeof input.ratelimitRefillInterval !== "number" ||
+          input.ratelimitRefillInterval <= 0
+        ) {
+          throw new TRPCError({
+            message: "Interval must be a positive integer",
+            code: "BAD_REQUEST",
+          });
+        }
+        ratelimitRefillInterval = input.ratelimitRefillInterval;
+      }
+      const key = await db.query.keys.findFirst({
+        where: (table, { eq }) => eq(table.id, input.keyId),
+        with: {
+          workspace: true,
+        },
+      });
+      if (!key) {
+        throw new TRPCError({ message: "key not found", code: "NOT_FOUND" });
+      }
+      if (key.workspace.tenantId !== ctx.tenant.id) {
+        throw new TRPCError({ message: "key not found", code: "NOT_FOUND" });
+      }
+      await db.transaction(async (tx) => {
+        await tx
+          .update(schema.keys)
+          .set({
+            ratelimitType,
+            ratelimitLimit,
+            ratelimitRefillRate,
+            ratelimitRefillInterval,
+          })
+          .where(eq(schema.keys.id, input.keyId));
+        await tx.insert(schema.auditLogs).values({
+          id: newId("auditLog"),
+          time: new Date(),
+          workspaceId: input.workspaceId,
+          apiId: null,
+          actorType: "user",
+          actorId: ctx.user.id,
+          event: "key.update",
+          description: `updated key ${input.keyId} rate limit`,
+          keyId: input.keyId,
+        });
+      });
+    }),
+  updateOwnerId: t.procedure
+    .use(auth)
+    .input(
+      z.object({
+        keyId: z.string(),
+        ownerId: z.string().nullish(),
```

Comment:
`nullish` means `string | null | undefined`, but we just want a string or null, right?
```suggestion
        ownerId: z.string().nullable(),
```

Reply:
correct this should not be undefined